ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLChatAgentRunner.java

```diff
@@ -470,27 +471,58 @@ private void runReAct(
+                                    ((ActionListener<Object>) nextStepListener)
+                                        .onResponse(
+                                            String
+                                                .format(
+                                                    Locale.ROOT,
+                                                    "Failed to run the tool %s with the error message %s.",
+                                                    action1,
+                                                    e.getMessage()
+                                                )
+                                        );
+                                });
+                                if (tools.get(action) instanceof MLModelTool) {
+                                    Map<String, String> llmToolTmpParameters = new HashMap<>();
+                                    llmToolTmpParameters.putAll(tmpParameters);
+                                    llmToolTmpParameters.putAll(toolSpecMap.get(action).getParameters());
+                                    // TODO: support tool parameter override : langauge_model_tool.prompt
+                                    llmToolTmpParameters.put(MLAgentExecutor.QUESTION, actionInput);
+                                    tools.get(action).run(llmToolTmpParameters, toolListener); // run tool
+                                } else {
+                                    tools.get(action).run(toolParams, toolListener); // run tool
+                                }
+                            } catch (Exception e) {
+                                ((ActionListener<Object>) nextStepListener)
+                                    .onResponse(
+                                        String
+                                            .format(
+                                                Locale.ROOT,
+                                                "Failed to run the tool %s with the error message %s.",
+                                                action,
+                                                e.getMessage()
+                                            )
+                                    );
                             }
                         } else {
-                            lastActionResult.set("Tool " + action + " can't work for input: " + actionInput);
+                            String res = String
+                                .format(Locale.ROOT, "Failed to run the tool %s due to wrong input %s.", action, actionInput);
+                            lastActionResult.set(res);
                             lastTool.set(action);
-                            String res = "Tool " + action + " can't work for input: " + actionInput;
                             ((ActionListener<Object>) nextStepListener).onResponse(res);
                         }
                     } else {
-                        lastTool.set(null);
+                        lastTool.set(action);
```

Comment:
what does `lastTool` used for? seem it never used. same for `lastToolListener` and `lastActionResult`

Reply:
Yeah, they are legacy code unused. We plan to refactor them afterwards. Anyway I removed them in this pr.